packages/babel-plugin-transform-optional-chaining/src/index.js

```diff
@@ -0,0 +1,160 @@
+  }
+
+  function isNodeOptional(node) {
+    return node.optional === true;
+  }
+
+  return {
+    visitor: {
+
+      AssignmentExpression(path, state) {
+        const { left } = path.node;
+
+        if (!isNodeOptional(left) || isOptionalTransformed(left)) {
+          return;
+        }
+
+        if (!state.optionalTemp) {
+          const id = path.scope.generateUidIdentifier();
+
+          state.optionalTemp = id;
+          path.scope.push({ id });
+        }
+
+        const { object, property } = left;
+
+        const isChainNil = t.BinaryExpression(
+          "!=",
+          createCondition(
+            state.optionalTemp,
+            object,
+            property,
+            nilIdentifier,
+          ),
+          nilIdentifier,
+        );
+
+        // FIXME(sven): if will be a ConditionalExpression for childs, only top level will be ifStatement
+        const remplacement = t.ifStatement(isChainNil, t.blockStatement([t.expressionStatement(path.node)]));
+
+        setOptionalTransformed(left);
+
+        path.traverse({
+          MemberExpression({ node }) {
+            setOptionalTransformed(node);
+          },
+        });
+
+        path.parentPath.replaceWith(remplacement);
+      },
+
+      UnaryExpression(path, state) {
+        const { operator, argument } = path.node;
+
+        if (operator !== "delete") {
+          return;
+        }
+
+        if (!isNodeOptional(argument) || isOptionalTransformed(argument)) {
+          return;
+        }
+
+        if (!state.optionalTemp) {
+          const id = path.scope.generateUidIdentifier();
+
+          state.optionalTemp = id;
+          path.scope.push({ id });
+        }
+
+        const { object, property } = argument;
+
+        const isChainNil = t.BinaryExpression(
+          "!=",
+          createCondition(
+            state.optionalTemp,
+            object,
+            property,
+            nilIdentifier,
+          ),
+          nilIdentifier,
+        );
+
+        const remplacement = t.ifStatement(isChainNil, t.blockStatement([t.expressionStatement(path.node)]));
+
+        setOptionalTransformed(argument);
+
+        path.parentPath.replaceWith(remplacement);
+      },
+
+      MemberExpression(path, state) {
+        if (!isNodeOptional(path.node) || isOptionalTransformed(path.node)) {
+          return;
+        }
+
+        const { object, property } = path.node;
+
+        if (!state.optionalTemp) {
+          const id = path.scope.generateUidIdentifier();
+
+          state.optionalTemp = id;
+          path.scope.push({ id });
+        }
+
+        if (t.isAssignmentExpression(path.parent)) {
+          return;
+        } else if (t.isUnaryExpression(path.parent)) {
+          return;
+        } else if (t.isCallExpression(path.parent)) {
+
+          const remplacement = createCondition(
+            state.optionalTemp,
+            object,
+            property,
+            t.callExpression(t.identifier("Function"), []),
```

Comment:
In the worst case it can be replaced with `t.functionExpression(null, [], t.blockStatement([]))`. This also avoids risk of shadowing, though, at the cost of 2 extra bytes when minified.